core/src/main/java/io/grpc/ManagedChannelBuilderProvider.java

```diff
@@ -0,0 +1,63 @@
+package io.grpc;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Provider of managed channels for transport agnostic consumption.
+ *
+ * <p>Implementations <em>should not</em> throw. If they do, it may interrupt class loading. If
+ * exceptions may reasonably occur for implementation-specific reasons, implementations should
+ * generally handle the exception gracefully and return {@code false} from {@link #isAvailable()}.
+ */
+@ExperimentalApi
+public abstract class ManagedChannelBuilderProvider {
+  private static final ManagedChannelBuilderProvider provider;
+
+  static {
+    ServiceLoader<ManagedChannelBuilderProvider> providers
+        = ServiceLoader.load(ManagedChannelBuilderProvider.class);
+    Iterator<ManagedChannelBuilderProvider> iter = providers.iterator();
+    List<ManagedChannelBuilderProvider> list = new ArrayList<ManagedChannelBuilderProvider>();
+    for (ManagedChannelBuilderProvider current : providers) {
+      if (!current.isAvailable()) {
+        continue;
+      }
+      list.add(current);
+    }
+    if (list.isEmpty()) {
+      provider = null;
+    } else {
+      provider = Collections.max(list, new Comparator<ManagedChannelBuilderProvider>() {
+        @Override
+        public int compare(ManagedChannelBuilderProvider f1, ManagedChannelBuilderProvider f2) {
+          return f1.priority() - f2.priority();
+        }
+      });
+    }
+  }
+
+  public static ManagedChannelBuilderProvider provider() {
+    return provider;
```

Comment:
So... ClassNotFoundException is not a RuntimeException, which means I'd need to turn it into something else at some point. That seems weak. NoClassDefFoundError is an Error and seems a weak semantic match.

So the XML provider stuff throws `FactoryConfigurationError` in case of failure. I created a `ProviderNotFoundException` that mimics a `RuntimeException` by the same name in java.nio.file. There is a `java.security.ProviderException` that looks reasonable too, but for some reason it has zero usages. It is a `RuntimeException` and its description seems apt:

> A runtime exception for Provider exceptions (such as misconfiguration errors or unrecoverable internal errors), which may be subclassed by Providers to throw specialized, provider-specific runtime errors.

I figure it is "good enough" now. I'd be happy to change it based on others preferences.